Source/Dafny/Compilers/Compiler.cs

```diff
@@ -4610,14 +4610,13 @@ private class GetterSetterLvalueImpl : ILvalue {
           wBody.Write(", {0}, ", expr is ForallExpr ? "true" : "false");
           var native = AsNativeType(e.BoundVars[i].Type);
           ConcreteSyntaxTree newWBody = CreateLambda(new List<Type> { bv.Type }, e.tok, new List<string> { IdName(bv) }, Type.Bool, wBody, untyped: true);
-          newWBody = EmitReturnExpr(newWBody);
+          newWBody = MaybeInjectSubsetConstraint(newWBody, inLetExprBody, e.BoundVars[i], e, true, e is ForallExpr);
```

Comment:
Very nice pointer, thank you very much, this lead further insights. The code that actually does the type test is unfortunately not located in `Compiler.cs`, but in every independent compiler `Compiler-CSharp.cs`,  `Compiler-go.cs`, etc. So it is not advisable to maintain this generic code in every compiler. I even believe the code of all these compilers should be ported back to `Compiler.cs` in a different effort.

Furthermore, your code helped me to find another soundness issue with respect to reference types:

```
trait Tr { }
class A extends Tr { }
class B extends Tr { }

function method test(x: Tr): bool
  requires x is A
{
  if x is B then 1/0 == 0 else true
}

method Main() {
  var a := new A;
  var b := new B;
  var s: set<Tr> := {a, b};
  var aa := forall a': A :: a' in s ==> test(a');
  assert(aa);
  print "ok";
}
```

In the case above, the verifier passes and
1) C#'s compilation complains that `error CS1503: Argument 1: cannot convert from 'System.Collections.Generic.IEnumerable<_module.Tr>' to 'System.Collections.Generic.IEnumer
able<_module.A>'`, which is something that @fabiomadge  is working on right now, so that could explain why nobody found this error before.
2) JavaScript it silently fails and "ok" is never printed;
3) Java: `Exception in thread "main" java.lang.ArithmeticException: BigInteger divide by zero`
4) Go: `Running... [Program halted] reflect: Call using *main.B as type *main.A` and nothing else displayed

The problem comes again from the fact that the type inference infers that a' is of type A and the bounds are a set of type Tr, and whereas there is some check for the foreach statement in every compiler, there is no check for compiling the forall because it compiles to a true mapping expression.

I'd say, all these errors would be automatically fixed if I implement [my right solution.](https://github.com/dafny-lang/dafny/issues/697#issuecomment-970703632).

For now, I suggest we merge this PR as soon as it is ready. This PR solves issues people encountered already, whereas the case above is less current. I could work on it on a second PR that works on type inference.